src/Essentials/src/FileSystem/FileSystem.ios.cs

```diff
@@ -158,68 +160,134 @@ protected override void Dispose(bool disposing)
 		}
 	}
 
-	class UIDocumentFileResult : FileResult
+	class UIDocumentFileResult : MediaFileResult
 	{
-		internal UIDocumentFileResult(NSUrl url)
-			: base()
+		UIDocument document;
+		NSUrl assetUrl;
+
+		internal UIDocumentFileResult(NSUrl assetUrl, string fileName)
 		{
-			var doc = new UIDocument(url);
-			FullPath = doc.FileUrl?.Path;
-			FileName = doc.LocalizedName ?? Path.GetFileName(FullPath);
+			this.assetUrl = assetUrl;
+			document = new UIDocument(assetUrl);
+			Extension = document.FileUrl.PathExtension!;
+			ContentType = GetMIMEType(document.FileType);
+			NameWithoutExtension = !string.IsNullOrWhiteSpace(fileName)
+				? Path.GetFileNameWithoutExtension(fileName)
+				: null;
+			Type = GetFileType(ContentType);
+			FileName = GetFileName(NameWithoutExtension, Extension);
 		}
 
 		internal override Task<Stream> PlatformOpenReadAsync()
-		{
-			Stream fileStream = File.OpenRead(FullPath);
+			=> Task.FromResult<Stream>(File.OpenRead(document.FileUrl.Path!));
 
-			return Task.FromResult(fileStream);
+		protected internal override void PlatformDispose()
+		{
+			document?.Dispose();
+			document = null;
+			assetUrl?.Dispose();
+			assetUrl = null;
+			base.PlatformDispose();
 		}
 	}
 
-	class UIImageFileResult : FileResult
+	class UIImageFileResult : MediaFileResult
 	{
-		readonly UIImage uiImage;
-		NSData data;
+		UIImage img;
+		NSDictionary metadata;
+		NSMutableData imgWithMetadata;
 
-		internal UIImageFileResult(UIImage image)
-			: base()
+		internal UIImageFileResult(UIImage img, NSDictionary metadata, string name)
 		{
-			uiImage = image;
-
-			FullPath = Guid.NewGuid().ToString() + FileExtensions.Png;
-			FileName = FullPath;
+			this.img = img;
+			this.metadata = metadata;
+			NameWithoutExtension = name;
+#pragma warning disable CA1422
+			ContentType = GetMIMEType(OldUTType.JPEG);
+			Extension = GetExtension(OldUTType.JPEG);
+#pragma warning restore CA1422
+			Type = GetFileType(ContentType);
+			FileName = GetFileName(NameWithoutExtension, Extension);
 		}
 
 		internal override Task<Stream> PlatformOpenReadAsync()
 		{
-			data ??= uiImage.AsPNG();
+			imgWithMetadata ??= GetImageWithMeta();
+			return Task.FromResult(imgWithMetadata?.AsStream());
```

Comment:
The compiler should have thrown a warning, we are returning a Task<Stream?> when the signature states is a Task<Stream>